### PR TITLE
feat: sync reporting viewset improvements

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
image

https://2u-internal.atlassian.net/browse/ENT-6464

tied to https://github.com/openedx/edx-enterprise/pull/1671

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: integrated_channels/api/v1/logs/serializers.py [hunk 1]
```diff
@@ -95,7 +95,7 @@ def get_sync_status(self, obj):
             sync_status = 'pending'
         elif not obj.status.isdigit() or int(obj.status) >= 400:
             sync_status = 'error'
-        elif int(obj.status) < 400:
+        elif int(obj.status) < 300:
```
Comment (alex-sheehan-edx) [inline integrated_channels/api/v1/logs/serializers.py:98]: technically 300's are failures

File: integrated_channels/api/v1/logs/views.py [hunk 2]
```diff
@@ -24,32 +26,65 @@
     LearnerDataTransmissionAudit,
 )
 
-from .serializers import ContentSyncStatusSerializer, GenericLearnerSyncStatusSerializer, LearnerSyncStatusSerializer
+from .serializers import ContentSyncStatusSerializer, LearnerSyncStatusSerializer
 
 LOGGER = logging.getLogger(__name__)
 
 
+class ReportingSyncStatusPagination(PageNumberPagination):
```
Comment (alex-sheehan-edx) [inline integrated_channels/api/v1/logs/views.py:34]: total number of pages is required for manual pagination so I added it to my own custom pagination

File: integrated_channels/api/v1/logs/views.py [hunk 3]
```diff
@@ -24,32 +26,65 @@
...
     Sync-status APIs for `ContentMetadataItemTransmission` items.
     """
     serializer_class = ContentSyncStatusSerializer
     permission_classes = (permissions.IsAuthenticated,)
     permission_required = 'enterprise.can_access_admin_dashboard'
-    pagination_class = PageNumberPagination
+    pagination_class = ReportingSyncStatusPagination
 
     def get_queryset(self):
-        enterprise_customer_uuid = self.kwargs.get('enterprise_customer_uuid', None)
```
Comment (alex-sheehan-edx) [inline integrated_channels/api/v1/logs/views.py:42]: specifying None is redundant

File: integrated_channels/api/v1/logs/views.py [hunk 4]
```diff
@@ -24,32 +25,83 @@
...
+        return Response(OrderedDict([
+            ('pages_count', self.page.paginator.num_pages),
+            ('count', self.page.paginator.count),
+            ('next', self.get_next_link()),
+            ('previous', self.get_previous_link()),
+            ('results', data)
+        ]))
+
+
+class InvalidSortByParamException(exceptions.APIException):
```
Comment (marlonkeating) [inline integrated_channels/api/v1/logs/views.py:46]: nit: These exceptions all have the same fields, maybe have a base class for them?

File: integrated_channels/api/v1/logs/views.py [hunk 5]
```diff
@@ -24,32 +25,83 @@
...
+        integrated_channel_code = self.kwargs.get('integrated_channel_code')
         channel_config_cls = EnterpriseCustomerPluginConfiguration.get_class_by_channel_code(integrated_channel_code)
         if channel_config_cls is None:
-            raise exceptions.ParseError("Invalid channel code.")
+            raise InvalidChannelCodeException
+
+        if sort_by_param := self.request.GET.get('sort_by'):
+            if sort_by_param.replace("-", "") not in ["sync_status", "sync_last_attempted_at"]:
+                if sort_by_param.replace("-", "") not in [field.name for field in ContentMetadataItemTransmission._meta.fields]:
+                    raise InvalidSortByParamException
```
Comment (marlonkeating) [inline integrated_channels/api/v1/logs/views.py:76]: Don't need to nest conditionals to test all the conditions
```suggestion
        if sort_by_param := self.request.GET.get('sort_by'):
            if sort_by_param.replace("-", "") not in ["sync_status", "sync_last_attempted_at"] + [field.name for field in ContentMetadataItemTransmission._meta.fields]:
                raise InvalidSortByParamException
```
If this breaks python line limits, you can always do
```suggestion
        if sort_by_param := self.request.GET.get('sort_by'):
            valid_sorts = ["sync_status", "sync_last_attempted_at"] + \
                [field.name for field in ContentMetadataItemTransmission._meta.fields]
            if sort_by_param.replace("-", "") not in valid_sorts:
                raise InvalidSortByParamException
```

File: integrated_channels/api/v1/logs/views.py [hunk 6]
```diff
@@ -24,32 +25,83 @@
...
+
+    def get_queryset(self):
+        self.validate()
+        content_metadata_transmission_audit_filter = {
+            "enterprise_customer_id": self.kwargs.get('enterprise_customer_uuid'),
+            "integrated_channel_code": self.kwargs.get('integrated_channel_code'),
+            "plugin_configuration_id": self.kwargs.get('plugin_configuration_id'),
+        }
```
Comment (marlonkeating) [inline integrated_channels/api/v1/logs/views.py:84]: Are we just picking out a few keys from a dictionary here?  If so, we could just do
```suggestion
        content_metadata_fields = ('enterprise_customer_uuid', 'integrated_channel_code', 'plugin_configuration_id')
        content_metadata_transmission_audit_filter = {k: self.kwargs.get(k) for k in content_metadata_fields}
```
If we do things like this a lot, it's a good candidate for making into a helper method, or using library like [pydash.pick](https://pydash.readthedocs.io/en/latest/api.html?highlight=pick#pydash.objects.pick)

File: integrated_channels/api/v1/logs/views.py [hunk 7]
```diff
@@ -67,15 +119,43 @@ def get_serializer_class(self):
...
+    def validate(self):
+        integrated_channel_code = self.kwargs.get('integrated_channel_code', None)
+        ThisLearnerClass = LearnerDataTransmissionAudit.get_completion_class_by_channel_code(integrated_channel_code)
+        if ThisLearnerClass is None:
+            raise InvalidChannelCodeException
+
+        if sort_by_param := self.request.GET.get('sort_by'):
+            if sort_by_param.replace("-", "") not in ["sync_status", "sync_last_attempted_at"]:
+                if sort_by_param.replace("-", "") not in [field.name for field in ThisLearnerClass._meta.fields]:
+                    raise InvalidSortByParamException
```
Comment (marlonkeating) [inline integrated_channels/api/v1/logs/views.py:131]: Almost identical code to line 73 above, looks like we can wrap into helper method.

File: integrated_channels/api/v1/logs/views.py (hunk 6 above)
Comment (alex-sheehan-edx) [reply integrated_channels/api/v1/logs/views.py:84]: to my knowledge we're not doing this very frequently, but I do like your suggestion!

**Review by marlonkeating — APPROVED** [review]

File: integrated_channels/api/v1/logs/views.py [hunk 8]
```diff
@@ -24,32 +25,87 @@
...
+    default_detail = "Invalid channel code."
+    default_code = "validation_error"
+
+
+def validate_sort_by_query_params(model_table, request):
+    expected_sort_by_filters = ["sync_status", "sync_last_attempted_at"] + \
+        [field.name for field in model_table._meta.fields]
+    if sort_by_param := request.GET.get('sort_by'):
+        if sort_by_param.replace("-", "") not in expected_sort_by_filters:
+            raise InvalidSortByParamException
```
[the remark this answers is not in the document]
Comment (alex-sheehan-edx) [reply integrated_channels/api/v1/logs/views.py:65]: nice catch

File: integrated_channels/api/v1/logs/views.py [hunk 9]
```diff
@@ -67,15 +123,35 @@ def get_serializer_class(self):
...
+
+        learner_data_fields = ('enterprise_customer_uuid', 'plugin_configuration_id')
+        learner_data_transmission_audit_filter = {field: self.kwargs.get(field) for field in learner_data_fields}
+
+        if self.request.GET.get('user_email'):
+            learner_data_transmission_audit_filter['user_email__contains'] = self.request.GET.get('user_email')
+        if self.request.GET.get('content_title'):
+            learner_data_transmission_audit_filter['content_title__contains'] = self.request.GET.get('content_title')
+
+        # Defaults to DESC status code to put failed transmission on top
```
Comment (kiram15) [inline integrated_channels/api/v1/logs/views.py:146]: ty!

File: integrated_channels/api/v1/logs/views.py [hunk 10]
```diff
@@ -67,15 +123,35 @@ def get_serializer_class(self):
...
+        content_sort_by = ('-status',)
+        if self.request.GET.get('sort_by'):
+            # Translate the serialized `sync_status` to its DB field equivelant `api_response_status_code`
```
Comment (kiram15) [inline integrated_channels/api/v1/logs/views.py:149]: nit: equivalent

Comment (alex-sheehan-edx) [reply integrated_channels/api/v1/logs/views.py:149]: lol good catch

**Review by kiram15 — APPROVED** [review]
couple quick things but looks great!

**Pull request merged**